pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,312 @@
+	manager.epDataStore[epData.id] = epData
+	for _, config := range manager.policyConfigs {
+		if config.policyConfigSelectsEndpoint(epData) {
+			err := manager.upsertPolicyEndpoint(config, epData)
+			if err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.K8sEndpointName: endpoint.Name,
+					logfields.K8sNamespace:    endpoint.Namespace,
+				}).Error("Error upserting pod mapping for pod.")
+				return
+			}
+		}
+	}
+}
+
+// OnDeleteEndpoint is the event handler for endpoint deletions.
+func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	id := types.NamespacedName{
+		Name:      endpoint.GetName(),
+		Namespace: endpoint.GetNamespace(),
+	}
+
+	var epData *endpointMetadata
+	var err error
+	if epData, err = manager.getEndpointMetadata(endpoint); err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sEndpointName: endpoint.Name,
+			logfields.K8sNamespace:    endpoint.Namespace,
+		}).Error("failed to get valid endpoint metadata, abort deleting endpoint mapping")
+		return
+	}
+
+	if policies, ok := manager.policyEndpoints[id]; ok {
+		for _, policy := range policies {
+			config := manager.policyConfigs[policy]
+			manager.deleteEgressMap(config, epData)
+		}
+		delete(manager.policyEndpoints, id)
+	}
+	delete(manager.epDataStore, id)
+}
+
+func (manager *Manager) getEndpointMetadata(endpoint *k8sTypes.CiliumEndpoint) (*endpointMetadata, error) {
+	var ipv4s []string
+	id := types.NamespacedName{
+		Name:      endpoint.GetName(),
+		Namespace: endpoint.GetNamespace(),
+	}
+
+	for _, pair := range endpoint.Networking.Addressing {
+		if pair.IPV4 != "" {
+			ipv4s = append(ipv4s, pair.IPV4)
+		}
+	}
+
+	data := &endpointMetadata{
+		ips:    ipv4s,
+		labels: labels.NewLabelsFromModel(endpoint.Identity.Labels).K8sStringMap(),
+		id:     id,
+	}
+
+	// Keep id to endpoint data mapping up to date
+	manager.epDataStore[id] = data
+	return data, nil
+}
+
+// isValidConfig validates the given policy config.
+func (manager *Manager) isValidConfig(config EgressPolicyConfig) error {
+	for _, policyConfig := range manager.policyConfigs {
+		if policyConfig.egressIP.String() == config.egressIP.String() {
+			return fmt.Errorf(
+				"CiliumEgressNatPolicy for egress IP %v already exists",
+				config.egressIP.String())
+
+		}
+	}
+	return nil
+}
+
+// upsertPolicyEndpoint updates or insert to endpoint policy mapping for given policy config and endpoints,
+// it also upserts egress map to keep in sync
+func (manager *Manager) upsertPolicyEndpoint(config *EgressPolicyConfig, epData *endpointMetadata) error {
+	if err := manager.updateEgressMap(epData.ips, config); err != nil {
+		return err
+	}
+
+	if endpointPolicies, ok := manager.policyEndpoints[epData.id]; ok {
+		for _, poID := range endpointPolicies {
+			log.Debug("Endpoint to policy mapping already exists.")
```

Comment:
I guess this debug statement should be on line 267 instead?

Reply:
fixed. thx